### ISSUE TITLE
Erroneous error message when Danger fails to find branch-able point in git tree

### ISSUE THREAD

**Issue by samdmarshall** [danger/danger#316]
When running CI today I encountered an error where Danger was reporting an error due to unable to find the branch point of the current working copy but ultimately reported the error as a problem with the DSL and that I should attempt to update the gem to see if that resolved anything. It seems to me that the error was that I was in a detached head state in the git repo and not an expected checkout state for the CI to correctly run. This caused some of the git commands that Danger uses to fail before reporting the DSL error. This is kinda tricky to get to reproduce since I think it was a bug in CircleCI running two different builds of the same branch concurrently. 

<details>
 <summary>Error Log</summary>
```
$ make ci
Checking for which... yes
Checking for cat... yes
Checking for cp... yes
Checking for touch... yes
Checking for find... yes
Checking for xargs... yes
Checking for rm... yes
Checking for python... yes
Checking for python3... yes
Checking for pip... yes
Checking for tox... yes
Checking for coverage... yes
Checking for gem... yes
Checking for danger... yes
============================
/home/ubuntu/virtualenvs/venv-2.7.10/bin/tox
GLOB sdist-make: /home/ubuntu/pyconfig/setup.py
py27 create: /home/ubuntu/pyconfig/.tox/py27
py27 installdeps: coverage
py27 inst: /home/ubuntu/pyconfig/.tox/dist/pyconfig-1.0.2.zip
py27 installed: coverage==4.1,pyconfig==1.0.2,pyparsing==2.1.5
py27 runtests: PYTHONHASHSEED='2212423871'
py27 runtests: commands[0] | coverage run --source=./pyconfig/ setup.py test
running test
running egg_info
writing entry points to pyconfig.egg-info/entry_points.txt
writing dependency_links to pyconfig.egg-info/dependency_links.txt
writing pyconfig.egg-info/PKG-INFO
writing top-level names to pyconfig.egg-info/top_level.txt
reading manifest file 'pyconfig.egg-info/SOURCES.txt'
writing manifest file 'pyconfig.egg-info/SOURCES.txt'
running build_ext
test_comments (tests.pyconfig_test.pyconfigTestCases) ... ok
test_conditionals (tests.pyconfig_test.pyconfigTestCases) ... ok
test_direct_assignment_automatic (tests.pyconfig_test.pyconfigTestCases) ... ok
test_direct_assignment_specific (tests.pyconfig_test.pyconfigTestCases) ... ok
test_export_with_keyword (tests.pyconfig_test.pyconfigTestCases) ... ok
test_export_without_keyword (tests.pyconfig_test.pyconfigTestCases) ... ok
test_flags_scheme_name (tests.pyconfig_test.pyconfigTestCases) ... ok
test_include (tests.pyconfig_test.pyconfigTestCases) ... ok
test_inherits (tests.pyconfig_test.pyconfigTestCases) ... ok
test_variable_substitution_with_use (tests.pyconfig_test.pyconfigTestCases) ... ok
test_variable_substitution_without_use (tests.pyconfig_test.pyconfigTestCases) ... ok

----------------------------------------------------------------------
Ran 11 tests in 0.054s

OK
py35 create: /home/ubuntu/pyconfig/.tox/py35
py35 installdeps: coverage
py35 inst: /home/ubuntu/pyconfig/.tox/dist/pyconfig-1.0.2.zip
py35 installed: coverage==4.1,pyconfig==1.0.2,pyparsing==2.1.5
py35 runtests: PYTHONHASHSEED='2212423871'
py35 runtests: commands[0] | coverage run --source=./pyconfig/ setup.py test
running test
running egg_info
writing dependency_links to pyconfig.egg-info/dependency_links.txt
writing top-level names to pyconfig.egg-info/top_level.txt
writing entry points to pyconfig.egg-info/entry_points.txt
writing pyconfig.egg-info/PKG-INFO
reading manifest file 'pyconfig.egg-info/SOURCES.txt'
writing manifest file 'pyconfig.egg-info/SOURCES.txt'
running build_ext
test_comments (tests.pyconfig_test.pyconfigTestCases) ... ok
test_conditionals (tests.pyconfig_test.pyconfigTestCases) ... ok
test_direct_assignment_automatic (tests.pyconfig_test.pyconfigTestCases) ... ok
test_direct_assignment_specific (tests.pyconfig_test.pyconfigTestCases) ... ok
test_export_with_keyword (tests.pyconfig_test.pyconfigTestCases) ... ok
test_export_without_keyword (tests.pyconfig_test.pyconfigTestCases) ... ok
test_flags_scheme_name (tests.pyconfig_test.pyconfigTestCases) ... ok
test_include (tests.pyconfig_test.pyconfigTestCases) ... ok
test_inherits (tests.pyconfig_test.pyconfigTestCases) ... ok
test_variable_substitution_with_use (tests.pyconfig_test.pyconfigTestCases) ... ok
test_variable_substitution_without_use (tests.pyconfig_test.pyconfigTestCases) ... ok

----------------------------------------------------------------------
Ran 11 tests in 0.059s

OK
___________________________________ summary ____________________________________
  py27: commands succeeded
  py35: commands succeeded
  congratulations :)
============================
/home/ubuntu/virtualenvs/venv-2.7.10/bin/coverage report
Name                                    Stmts   Miss  Cover
-----------------------------------------------------------
pyconfig/Deserializer/Comment.py            0      0   100%
pyconfig/Deserializer/Include.py            0      0   100%
pyconfig/Deserializer/KeyValue.py           0      0   100%
pyconfig/Deserializer/Resolver.py           0      0   100%
pyconfig/Deserializer/XCLineItem.py         0      0   100%
pyconfig/Deserializer/__init__.py           0      0   100%
pyconfig/Deserializer/xcconfig.py           0      0   100%
pyconfig/Graph/Grapher.py                  21      4    81%
pyconfig/Graph/Searcher.py                 20      4    80%
pyconfig/Graph/__init__.py                  1      0   100%
pyconfig/Helpers/Logger.py                 23      0   100%
pyconfig/Helpers/OrderedDictionary.py      16      0   100%
pyconfig/Helpers/__init__.py                0      0   100%
pyconfig/Interpreter/Consumer.py           15      0   100%
pyconfig/Interpreter/Dependent.py          48      5    90%
pyconfig/Interpreter/LangParser.py         24      0   100%
pyconfig/Interpreter/__init__.py            1      0   100%
pyconfig/Keyword/BaseKeyword.py             3      1    67%
pyconfig/Keyword/Constants.py              12      0   100%
pyconfig/Keyword/ExportKeyword.py          13      2    85%
pyconfig/Keyword/IncludeKeyword.py         16      2    88%
pyconfig/Keyword/Resolver.py               13      0   100%
pyconfig/Keyword/SettingKeyword.py        100      7    93%
pyconfig/Keyword/Words.py                  11      0   100%
pyconfig/Keyword/__init__.py                3      0   100%
pyconfig/Serializer/Serializer.py          36      0   100%
pyconfig/Serializer/__init__.py             0      0   100%
pyconfig/__init__.py                        3      0   100%
pyconfig/main.py                           26      0   100%
pyconfig/version.py                         2      0   100%
pyconfig/version_info.py                    2      0   100%
-----------------------------------------------------------
TOTAL                                     409     25    94%
/home/ubuntu/virtualenvs/venv-2.7.10/bin/coverage html
/bin/cp -r ./htmlcov /tmp/circle-artifacts.S00S6Zl
============================
/opt/circleci/.rvm/gems/ruby-2.1.3/bin/danger
fatal: Not a valid branch point: 'f19e9405a69364fefba62f911fcacde8e0f78e29'.

[!] Invalid `Dangerfile` file: git '--git-dir=/home/ubuntu/pyconfig/.git' '--work-tree=/home/ubuntu/pyconfig' diff '--numstat' 'danger_base' 'danger_head'  2>&1:fatal: ambiguous argument 'danger_head': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'. Updating the Danger gem might fix the issue.

 #  from Dangerfile:9
 #  -------------------------------------------
 #  # Warn when there is a big PR
 >  warn("Big PR") if lines_of_code > 75
 #  
 #  -------------------------------------------
 (Danger::DSLError)
	/opt/circleci/.rvm/gems/ruby-2.1.3/gems/git-1.3.0/lib/git/lib.rb:937:in `command'
	/opt/circleci/.rvm/gems/ruby-2.1.3/gems/git-1.3.0/lib/git/lib.rb:861:in `command_lines'
	/opt/circleci/.rvm/gems/ruby-2.1.3/gems/git-1.3.0/lib/git/lib.rb:362:in `diff_stats'
	/opt/circleci/.rvm/gems/ruby-2.1.3/gems/git-1.3.0/lib/git/diff.rb:113:in `cache_stats'
	/opt/circleci/.rvm/gems/ruby-2.1.3/gems/git-1.3.0/lib/git/diff.rb:34:in `lines'
	/opt/circleci/.rvm/gems/ruby-2.1.3/gems/danger-0.8.5/lib/danger/danger_core/plugins/dangerfile_git_plugin.rb:73:in `lines_of_code'
	/opt/circleci/.rvm/gems/ruby-2.1.3/gems/danger-0.8.5/lib/danger/danger_core/dangerfile.rb:61:in `block in method_missing'
	/opt/circleci/.rvm/gems/ruby-2.1.3/gems/danger-0.8.5/lib/danger/danger_core/dangerfile.rb:59:in `each'
	/opt/circleci/.rvm/gems/ruby-2.1.3/gems/danger-0.8.5/lib/danger/danger_core/dangerfile.rb:59:in `method_missing'
	Dangerfile:9:in `block in parse'
	/opt/circleci/.rvm/gems/ruby-2.1.3/gems/danger-0.8.5/lib/danger/danger_core/dangerfile.rb:192:in `eval'
	/opt/circleci/.rvm/gems/ruby-2.1.3/gems/danger-0.8.5/lib/danger/danger_core/dangerfile.rb:192:in `block in parse'
	/opt/circleci/.rvm/gems/ruby-2.1.3/gems/danger-0.8.5/lib/danger/danger_core/dangerfile.rb:188:in `instance_eval'
	/opt/circleci/.rvm/gems/ruby-2.1.3/gems/danger-0.8.5/lib/danger/danger_core/dangerfile.rb:188:in `parse'
	/opt/circleci/.rvm/gems/ruby-2.1.3/gems/danger-0.8.5/lib/danger/commands/runner.rb:60:in `run'
make: *** [danger] Error 1
```
</details>

**Comment by orta** [comment]
I wonder if there's a way of requesting a commit range from a Git repo, woe could do that beforehand to ensure the before/after range is 👍

**Comment by samdmarshall** [comment]
I saw this again last night, looks like if you push any commits to a branch that CircleCI is watching while it is already running a session it will spawn a new session for each additional commit that was made. This results in all but the most recent commit failing (in my experience). I guess this can be added to the list of things that is kinda broken with CircleCI?

**Comment by orta** [comment]
So the case here is this?

* Push a few commits, make a PR
* While that is building, you push new commits
* GitHub's PR API now references git commits after the commit range
* CI build gets around to running Danger, which uses the GitHub API due to workarounds
* Danger bails, because those commits don't exist in the current scop

**Comment by orta** [comment]
If so, we _could_ make a case for either:

1. Checking for the PR's head commit existing, and if not, using the current `:head`.
2. Fetching the latest commits.

**Comment by segiddins** [comment]
Is there any reason we can't use `git merge-point HEAD $TARGET_BRANCH` to figure out the commit range?

**Comment by orta** [comment]
Mainly because I've never heard of it 👍 - it could be what we're looking for

**Comment by jamtur01** [comment]
See also #160

**Comment by orta** [comment]
This now _should_ be fixed in 3.2.1 (which ensures all commits exist) - if it wasn't fixed in 2.x ( whenever we introduced `merge-point`'s )
